dhall/src/Dhall/Normalize.hs

```diff
@@ -211,8 +211,13 @@ normalizeWithM ctx e0 = loop (Syntax.denote e0)
                         strict =       strictLoop (fromIntegral n0 :: Integer)
                         lazy   = loop (  lazyLoop (fromIntegral n0 :: Integer))
 
-                        strictLoop 0 = loop zero
-                        strictLoop !n = App succ' <$> strictLoop (n - 1) >>= loop
+                        strictLoop !n = strictLoopShortcut n (loop zero)
+                        strictLoopShortcut 0 !res = res
+                        strictLoopShortcut !n !res = do
+                            x <- res
+                            let next_res = App succ' <$> res >>= loop
+                            y <- next_res
+                            if judgmentallyEqual x y then res else strictLoopShortcut (n - 1) next_res
```

Comment:
I think you can avoid evaluating `rest` and `next_res` twice by structuring it like this:

```suggestion
                        strictLoop !n = do
                            z <- loop zero
                            strictLoopShortcut n z

                        strictLoopShortcut 0 !previous = pure previous
                        strictLoopShortcut !n !previous = do
                            current <- loop (App succ' previous)
                            if judgmentallyEqual previous current
                                then pure previous
                                else strictLoopShortcut (n - 1) current
```